speck/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+//! Pure Rust implementation of the [Speck] block cipher.
+//!
+//! [Speck]: https://en.wikipedia.org/wiki/Speck_(cipher)
+
+#![no_std]
+#![doc(
+    html_logo_url = "https://raw.githubusercontent.com/RustCrypto/media/26acc39f/logo.svg",
+    html_favicon_url = "https://raw.githubusercontent.com/RustCrypto/media/26acc39f/logo.svg",
+    html_root_url = "https://docs.rs/speck/0.0.1"
+)]
+#![forbid(unsafe_code)]
+#![warn(missing_docs, rust_2018_idioms)]
+
+pub use cipher;
+
+use cipher::{
+    consts::{U12, U16, U18, U24, U32, U4, U6, U8, U9},
+    AlgorithmName, BlockCipher, Key, KeyInit, KeySizeUser,
+};
+use core::fmt;
+
+macro_rules! define_speck_impl {
+    (
+        $name:ident,
+        $block_size:ty,
+        $key_size:ty,
+        $word_type:ty,
+        $n:literal,
+        $m:literal,
+        $alpha:literal,
+        $beta:literal,
+        $mask:literal,
+        $rounds:literal,
+        $doc:expr $(,)?
+    ) => {
+        #[doc=$doc]
+        #[doc = "block cipher"]
+        #[derive(Clone)]
+        pub struct $name {
+            k: [$word_type; $rounds],
+        }
+
+        impl $name {
+            #[inline]
+            fn from_be_bytes(bytes: [u8; $n / 8]) -> $word_type {
+                // We can't use $word_type::from_be_bytes here because the word size might be different from the word type.
```

Comment:
Wouldn't it be better to simply copy bytes into `[u8; size_of::<$word_type>()]` and call `from_be_bytes` on it? I think it also could be a bit more efficient. You also could pass `bytes` as `&[u8]` thus removing the try conversion step (compiler should eliminate all bound checks since all sizes are known at compile time).

Same for the `to_be_bytes` method.

Reply:
Yes, but we would be making an additional array. Are you sure this would be more efficient?

Reply:
Yes. Handling of such arrays gets heavily optimized by compiler. See generated assembly: https://rust.godbolt.org/z/MME5f6Wcc

Reply:
Thank you!